### PR TITLE
CI - SBPF program build script

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
The build script is very slow and rebuilds everything always even if nothing changed.
Also, the host and guest targets are build for both the test programs and the tests themselves, which is unnecessary.

#### Summary of Changes
- Uses `cargo build --workspace` directly instead of going through `cargo-build-sbf` individually
- Removes the Rust build script
- Adds a Makefile to build C, build Rust, and test both
- Uses cargo strip instead of the strip script

#### Improvement
Rebuild one or fresh build all test program before:
real    2m9.785s
user    1m35.221s
sys     0m34.474s

Fresh build all test programs after:
real    0m31.475s
user    1m52.219s
sys     0m19.844s

Rebuild one test program after:
real    0m0.639s
user    0m0.438s
sys     0m0.211s

Meaning for rebuilding a single test program we get a speedup of 129.785s / 0.639s ~ 200x or need 1 - 0.639s / 129.785s ~ 99.5% less time.

#### Usage
To build and run all tests do: `make test`
And to run a specific test do e.g.: `make test TEST_ARGS=realloc_invoke`

### CODE REVIEW

**Comment by alessandrod** [comment]
> #### Problem
> 
> The build script is very slow and rebuilds everything always even if nothing changed. Also a bunch of tests are effectively ignored because they are behind the `cfg(feature = "test-bpf")` which is never set.

test-bpf is set by cargo-test-sbf I think?


> #### Summary of Changes

In addition to saying _what_, could you also briefly say _why_ you changed things?

>     * Uses `cargo build --workspace` directly instead of going through `cargo-build-sbf` individually

The goal here is to avoid paying the cost of calling cargo-build-sbf a number of times, which is slow, right?

>     * Removes the Rust build script

This is self explanatory 😍

>     * Adds a Makefile to build C, build Rust, and test both
> 
>     * Fixes the names of "sibling_inner_instruction", "sibling_instruction", "big_mod_exp"
> 
>     * Moves the tests ignored through `cfg(feature = "test-bpf")` into tests/programs.rs instead: "mem", "remaining_compute_units", "sanity", "simulation", "sysvar"

Could this be done in a separate PR for ease of reviewing? And also, could you explain why? programs.rs is already huge.

>     * Fixes these previously ignored tests as they did not get exercised in a while

Why were these not being exercised? Are we not running cargo-test-sbf in CI?

>     * Splits the following into a separate "dep" crate each: "invoke", "invoked", "mem", "realloc", "realloc-invoke"
>     * Adjusts the `crate-type` in these accordingly and removes "program" feature cfg

Could you explain why this was needed?

>     * Uses cargo strip instead of the strip script

Yay

**Comment by Lichtso** [comment]
> test-bpf is set by cargo-test-sbf I think?

You are right, there is an extra shell script loop that runs them. The question becomes why some of them fail if I run them manually without the `#[!cfg(feature = "test-bpf")]`. Might be I am missing some env vars.

> Could this be done in a separate PR for ease of reviewing?

Already planned on doing so, this PR is just to test if I can get all CI to pass and not removing tests.

> And also, could you explain why? programs.rs is already huge.

Agreed, and I think we should split programs.rs into multiple files. I thought about moving them all in their respective crate tests directories, but that does not work for all the C tests. Additionally, with the crate tests directories the issue is that these then have dev-dependencies and these get build for the host arch in release mode (which is never used) because workspace dependencies can not distinguish between dependencies and dev-dependencies. So I decided to go for the opposite direction and move everything into the workspace global tests directory. That way its building can be separated and we build all the test programs for the guest only and all the tests for the host only, instead of building both for both.

> Could you explain why this was needed?

The splitting is necessary because when building the entire `--workspace` it can not be done with features as every crate can only be build with one feature set. So the previous strategy of splitting crates internally with cfg features only works for individual builds. Now we have to split them explicitly.

File: programs/sbf/Cargo.toml [hunk 1]
```diff
@@ -69,14 +69,14 @@ homepage = { workspace = true }
 license = { workspace = true }
 edition = { workspace = true }
 
+[profile.release]
+strip = true
```
Comment (alessandrod) [inline programs/sbf/Cargo.toml:75]: I think we should have a comment here that explains why it's needed

**Review by alessandrod — COMMENTED** [review]
Looks so good, thanks so much for doing this!

It would be good to have a way to pass arguments to cargo test, so that you can filter the tests to run?

File: programs/sbf/Makefile [hunk 2]
```diff
@@ -0,0 +1,13 @@
+SBF_SDK_PATH := ../../sdk/sbf
+SRC_DIR := c/src
+OUT_DIR := target/sbf-solana-solana/release
+
+test: rust all
+	SBF_OUT_DIR=$(OUT_DIR) cargo test --features="sbf_rust,sbf_c" $(test)
```
Comment (alessandrod) [inline programs/sbf/Makefile]: nit: I think this should be `$(TEST_ARGS)`

The convention is to have variables be uppercase, and the variable can contain more than just one test name. For example I would run `make test TEST_ARGS="sometest -- --nocapture"`

**Review by alessandrod — APPROVED** [review]

**Pull request merged**